array_api_compat/torch/linalg.py

```diff
@@ -22,6 +22,34 @@ def cross(x1: array, x2: array, /, *, axis: int = -1) -> array:
     x1, x2 = _fix_promotion(x1, x2, only_scalar=False)
     return torch_linalg.cross(x1, x2, dim=axis)
 
-__all__ = linalg_all + ['outer', 'trace', 'matrix_transpose', 'tensordot']
+def vecdot(x1: array, x2: array, /, *, axis: int = -1, **kwargs) -> array:
+    from ._aliases import isdtype
+
+    x1, x2 = _fix_promotion(x1, x2, only_scalar=False)
+
+    # torch.linalg.vecdot doesn't support integer dtypes
+    if isdtype(x1.dtype, 'integral') or isdtype(x2.dtype, 'integral'):
+        if kwargs:
+            raise RuntimeError("vecdot kwargs not supported for integral dtypes")
+        ndim = max(x1.ndim, x2.ndim)
+        x1_shape = (1,)*(ndim - x1.ndim) + tuple(x1.shape)
+        x2_shape = (1,)*(ndim - x2.ndim) + tuple(x2.shape)
+        if x1_shape[axis] != x2_shape[axis]:
+            raise ValueError("x1 and x2 must have the same size along the given axis")
```

Comment:
Now that I think about it, I wonder if this sort of thing should already be disallowed by the spec. https://data-apis.org/array-api/latest/API_specification/generated/array_api.vecdot.html#array_api.vecdot

I've been working on the assumption that axis refers to the dimension after broadcasting ("Must be an integer on the interval [-N, N), where N is the rank (number of dimensions) of the shape determined according to [Broadcasting](https://data-apis.org/array-api/latest/API_specification/broadcasting.html#broadcasting)."). But it also says "The contracted axis (dimension) must not be broadcasted." 

I had been interpreting that as meaning you shouldn't allow something like `vecdot(empty((3, 3)), empty((1, 3)), axis=0)`. But I suppose it could also be taken to mean that broadcasting shouldn't "broadcast up" to the contracted dimension either. Something more along the lines of

```py
if axis >= 0:
    ndim = max(x.ndim for x in inputs)
    if any(axis < ndim - x.ndim for x in inputs):
        raise ValueError("Contracted axis cannot be broadcast")
```

(e.g., `vecdot(empty((1, 2, 3, 4, 5)), empty((3, 4, 5)), axis=2)` is fine but `vecdot(empty((1, 2, 3, 4, 5)), empty((3, 4, 5)), axis=0)` is not)

Reply:
I think disallowing both (not broadcast along the reduced dimension and ask for the axis to be well defined in the sense you just described) would be the safer thing to do, as otherwise you end up with these funny constructions, which is not great.

Reply:
This is an issue in all the other implementations too, and arguably the spec as well. I'm going to deal with it in a separate PR.